File: crates/contracts/src/market_data.rs
//! Normalized market data bundle contract (market-data-bundle.schema.json).

use serde::{Deserialize, Serialize};

use crate::Decimal;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Candle {
    pub ts: String,
    pub open: Decimal,
    pub high: Decimal,
    pub low: Decimal,
    pub close: Decimal,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub volume: Option<Decimal>,
}

fn default_quote() -> String {
    "USD".to_string()
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct CandleSeries {
    pub venue: String,
    pub symbol: String,
    #[serde(default = "default_quote")]
    pub quote: String,
    #[serde(default)]
    pub points: Vec<Candle>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FundingPoint {
    pub ts: String,
    pub rate: Decimal,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct FundingSeries {
    pub venue: String,
    pub symbol: String,
    #[serde(default)]
    pub points: Vec<FundingPoint>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GasPoint {
    pub ts: String,
    pub gas_usd: Decimal,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct GasSeries {
    pub chain: String,
    #[serde(default)]
    pub points: Vec<GasPoint>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct YieldPoint {
    pub ts: String,
    pub apr: Decimal,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct YieldSeries {
    pub protocol: String,
    pub asset: String,
    pub chain: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pool: Option<String>,
    #[serde(default)]
    pub points: Vec<YieldPoint>,
}

/// An AMM pool's reserves at a tick — `reserve_base` units of the base token and
/// `reserve_quote` units of the quote (USD/stable). Constant-product price impact
/// is computed from these (#40).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LiquidityPoint {
    pub ts: String,
    pub reserve_base: Decimal,
    pub reserve_quote: Decimal,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct LiquiditySeries {
    pub venue: String,
    pub symbol: String,
    #[serde(default)]
    pub points: Vec<LiquidityPoint>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Coverage {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub start: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub end: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub complete: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Provider {
    pub name: String,
    pub kind: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub coverage: Option<Coverage>,
    /// Data provenance: `native` (the venue's own price/feed), `reference` (a
    /// proxy, e.g. a CEX price stored under another venue), or `derived`. Lets
    /// results tell a Hyperliquid mark / Base-DEX price from a Binance reference.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provenance: Option<String>,
    /// The series this provider covers, for per-series providers (else None).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub venue: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub symbol: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct MarketDataBundle {
    #[serde(default = "default_bundle_schema_version")]
    pub schema_version: String,
    pub interval: String,
    pub start: String,
    pub end: String,
    #[serde(default)]
    pub candles: Vec<CandleSeries>,
    #[serde(default)]
    pub funding: Vec<FundingSeries>,
    #[serde(default)]
    pub gas: Vec<GasSeries>,
    #[serde(default)]
    pub yields: Vec<YieldSeries>,
    #[serde(default)]
    pub liquidity: Vec<LiquiditySeries>,
    #[serde(default)]
    pub providers: Vec<Provider>,
    #[serde(default)]
    pub warnings: Vec<String>,
}

fn default_bundle_schema_version() -> String {
    "catalyst.backtest.market_data_bundle.v1".to_string()
}

File: packages/contracts/src/catalyst_contracts/market_data.py
"""Normalized market data bundle contract models (market-data-bundle.schema.json)."""

from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from ._base import Decimal, StrictModel
from .request import Interval


class Candle(StrictModel):
    ts: datetime
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    volume: Decimal | None = None


class CandleSeries(StrictModel):
    venue: str
    symbol: str
    quote: str = "USD"
    points: list[Candle] = Field(default_factory=list)


class FundingPoint(StrictModel):
    ts: datetime
    rate: Decimal


class FundingSeries(StrictModel):
    venue: str
    symbol: str
    points: list[FundingPoint] = Field(default_factory=list)


class GasPoint(StrictModel):
    ts: datetime
    gas_usd: Decimal


class GasSeries(StrictModel):
    chain: str
    points: list[GasPoint] = Field(default_factory=list)


class YieldPoint(StrictModel):
    ts: datetime
    apr: Decimal


class YieldSeries(StrictModel):
    protocol: str
    asset: str
    chain: str
    pool: str | None = None
    points: list[YieldPoint] = Field(default_factory=list)


class Coverage(StrictModel):
    start: datetime | None = None
    end: datetime | None = None
    complete: bool | None = None


class Provider(StrictModel):
    name: str
    kind: Literal["candles", "funding", "gas", "yields"]
    coverage: Coverage | None = None
    provenance: Literal["native", "reference", "derived"] | None = None
    venue: str | None = None
    symbol: str | None = None


class MarketDataBundle(StrictModel):
    schema_version: str = "catalyst.backtest.market_data_bundle.v1"
    interval: Interval
    start: datetime
    end: datetime
    candles: list[CandleSeries] = Field(default_factory=list)
    funding: list[FundingSeries] = Field(default_factory=list)
    gas: list[GasSeries] = Field(default_factory=list)
    yields: list[YieldSeries] = Field(default_factory=list)
    providers: list[Provider] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


__all__ = [
    "MarketDataBundle",
    "CandleSeries",
    "Candle",
    "FundingSeries",
    "GasSeries",
    "YieldSeries",
    "Provider",
    "Coverage",
]
